Connecting library application to SQL Server using Windows forms (#1767)
section-engineering-education/engineering-education: articles/connecting-library-application-to-sql-server-using-windows-forms/index.md

# Connecting-library-application-to-SQL-Server-using-Windows-forms

### Introduction

> windows forms is one of the best tools used in making applications.
> With windows forms, you can make a variety of applications, from a mini calculator to a library application that connects to a database. In this tutorial, we will connect a library application using Windows forms to a database using SQL Server.
> 
> In this tutorial, we will use a windows forms project and work on it, we will add new forms that will add edit & delete books from the library, we will also manage the authors, the stores & the publishers stored in the database. We will display the information of each attribute & refresh it each time a change accrues.

Reviewer: ```suggestion
Windows forms are one of the best tools used in making applications. With windows forms, you can make a variety of applications, from a mini calculator to a library application that uses a database.

In this tutorial, we will connect a library application to a SQL database server using Windows forms. We will use a windows forms project to add new forms, edit or delete from the library. Also, we will manage the authors, the stores & the publishers in the database, and display the information of each attribute & refresh it each time, a change accrues.

```

Reviewer: Please explain what a `Windows form` is. Add them in the first para. Give a 1 -2 line intro.

If possible, add a link.

### Prerequisites

Before we begin, it would help you as the reader to have the following:

> - A basic understanding of C# programming language.

Reviewer: ```suggestion
- A basic understanding of the C# programming language.

```

- A basic understanding of Windows forms.
- Visual Studio installed on your system.
- SQL Server installed on your system.

> If you don’t have Visual Studio installed on your computer, you can check this article on how to set up the C# environment in Visual Studio [here](https://www.geeksforgeeks.org/setting-environment-c-sharp/), and for SQL Server you can download it from [here](https://www.microsoft.com/en-us/sql-server/sql-server-downloads). If you are new to Windows forms you can check this tutorial that would help you understand The basics of it [here](https://www.section.io/engineering-education/getting-started-with-windows-forms-using-c-sharp/).

Reviewer: ```suggestion
#### Installation guide
- To install Visual Studio and setup the work environment, refer [this](https://www.geeksforgeeks.org/setting-environment-c-sharp/) article on how to set up the C# environment in Visual Studio.
- To install SQL Server, you can download it from [here](https://www.microsoft.com/en-us/sql-server/sql-server-downloads).
- If you are new to Windows forms, you can check [this](/engineering-education/getting-started-with-windows-forms-using-c-sharp/) tutorial which would help you understand the basic concepts.
```

Reviewer: Do not clutter too many information as a paragraph. Instead, list them out as points.

Also, while referring any other Section article, you need not type out the full URL. You can use `/engineering-education/getting-started-with-windows-forms-using-c-sharp/`

> 

Reviewer: Explain your objective about what you're going to build

> 

Reviewer: Add a `Table of contents` section.

### Creating the pubs database

> 

Reviewer: Could you be more specific on what "pubs" mean? And, also add why you need to create it

First of all, you should create the pubs database Query in SQL server so it can be later connected to the application.
You can find the pubs database [here](https://github.com/microsoft/sql-server-samples/blob/master/samples/databases/northwind-pubs/instpubs.sql).

> After copying the database paste it in a new Query and execute it, you should be able to see the name of the database "pubs" to the left of your screen.

Reviewer: If "pubs" is some sort of variable or unique name, you can put them into ` `` ` like `Pubs`. Also, capitalize the first letter if needed.

I believe "Pubs" is the short hand form of "Publishers". Please correct me, if I'm wrong.

> After copying the database paste it in a new Query and execute it, you should be able to see the name of the database "pubs" to the left of your screen.

Reviewer: `After copying the database`.

Did you mean `database query`?

![CreatingPubs](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/CreatingPubs.png).

### Application main form

We will use a project that has a main form & add another 3 forms that would add edit & delete books from the database.
You can download the project from [here](https://github.com/mohamedgh16/Library-app-code-in-windowsforms).

After opening the project first we will work on the main form, As you can see it has 3 buttons, each button for a different operation.

![Mainform](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/Mainform.png).

> Add a refresh button and 3 more tabs next to the Title tab to use later, then add 4 more columns to the datagridview to show the data of the library.

Reviewer: `datagridview`.

What is "datagridview"? Does it come as one single word? If not, split them out

Also, don't forget to change the colors of the forms & the buttons to whatever suits your taste.

It should look like this ![Afterformwork](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/Afterformwork.png).

### Database connection
In order to connect to SQL server in c#, we will be
```c#
using System.Data.SqlClient; 
```          
The following code snippet will connect your computer to the database, but note that you need to put your own computer name instead of (HP-PAVILION).                                          

> ```c#

Reviewer: Add comments for all the code snippets

static SqlConnection connectionString = new SqlConnection(@"Data Source=HP-PAVILION;Initial Catalog=pubs;Integrated Security=True");
```
Each Select Query will be written in a string and get converted to SQL server then back to use it somewhere inside the code. 
```c#
 public static DataTable dataAdapterSelect(string sqlQuery)
        {
            SqlDataAdapter DataAdapter = new SqlDataAdapter(sqlQuery, connectionString);
            DataTable dt = new DataTable();
            DataAdapter.Fill(dt);
            return dt;

>         }

Reviewer: Add inline code comments

```
The following method will be used to Add Edit & delete the information from the database.
```c#
public static void sqlCommandQueryReader(string sqlQuery)
        {
            SqlCommand myCommand = new SqlCommand(sqlQuery, connectionString);
            myCommand.Connection.Open();
            SqlDataReader dr;
            dr = myCommand.ExecuteReader();
            while (dr.Read())
            {
                Console.WriteLine(dr[0]);
                Console.WriteLine(dr[1]);
            }
            myCommand.Connection.Close();

>         }

Reviewer: Add inline code comments

```

 
The following code will use the dataAdapterSelect method to convert the Select string to a Query and display it inside the datagridview.
```c#
 string sqlQuery = "SELECT title_id,title,type,pub_name,price,ytd_sales FROM titles inner join publishers ON titles.pub_id = publishers.pub_id";

            DataTable dt = DataBaseConnection.dataAdapterSelect(sqlQuery);
           
            foreach (DataRow dr in dt.Rows)
            {
                dataGridView1.Rows.Add(dr["title_id"], dr["pub_name"], dr["price"], dr["ytd_sales"], dr["title"],dr["type"]);

>             }

Reviewer: Format the code

Reviewer: Add inline code comments

```

### Add form 

> Adding a new book will require the title id title name title type publisher id publisher date & the price of the book.

Reviewer: Add these variables within ` `` `

> Adding a new book will require the title id title name title type publisher id publisher date & the price of the book.

Reviewer: ```suggestion
Adding a new book will require the `titleID`, `titleName`, `titletype`, `pubid`, `pubdate`, and `price` of the book.

```

Reviewer: Make these kind of changes throughout the article

We will request the information from the user and insert them into the database using the sqlCommandQueryReader method.
![Addform](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/Addform.png).

```c#
  private void addButton_Click(object sender, EventArgs e)
        {   string titleID = titleIDTextBox.Text;
            string titleName = titleNameTextBox.Text;
            string titletype = Titletype.Text;
            string pubid = pubbid.Text;
            string pricee = price.Text;
            string pubbdate = pubdate.Text;

            string insertt = "insert into titles(title_id,title,type,pub_id,price,pubdate) values('" + titleID + 
            "','" + titleName + "','" + titletype + "','" + pubid + "','" + pricee + "','" + pubbdate + "');";

            DataBaseConnection.sqlCommandQueryReader(insertt);

>             MessageBox.Show("Information inserted!");}

Reviewer: Add inline code comments

        {            
            string dell = Del.Text;
            string delete1 = "delete from roysched where roysched.title_id='"+dell+"'; delete from sales where sales.title_id='"+dell+"';delete from titles where titles.title_id='"+dell+"';";
            if (Del.Text.Length != 0)
            {
                DataBaseConnection.sqlCommandQueryReader(delete1);
                MessageBox.Show("Title has been deleted!");
            }
            else
                MessageBox.Show("enter a title id!");
        }
```
### Refresh button

The refresh button will simply display the information from the database and will be needed to be clicked after any operation to keep the information updated.

```c#
 private void refresh_Click(object sender, EventArgs e)
        { dataGridView1.Rows.Clear();
          string sqlQuery = "SELECT title_id,title,type,pub_name,price,ytd_sales FROM titles inner join publishers ON titles.pub_id = publishers.pub_id";
          DataTable dt = DataBaseConnection.dataAdapterSelect(sqlQuery);
          foreach (DataRow dr in dt.Rows)
        { dataGridView1.Rows.Add(dr["title_id"], dr["pub_name"], dr["price"], dr["ytd_sales"], dr["title"], dr["type"]);}}

> ```

Reviewer: Format the code snippet neatly.

Reviewer: Add inline code comments

### Authors tab

The author's tab will display the id first name phone address city & count of each book the author wrote.

![Authors](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/Authors.png).

```c#
 string sqlQuery2 = "select * from authors order by au_fname asc";
            DataTable dt2 = DataBaseConnection.dataAdapterSelect(sqlQuery2);
            foreach (DataRow dr in dt2.Rows)
            {dataGridView2.Rows.Add(dr["au_id"], dr["au_fname"], dr["phone"], dr["address"], dr["city"]);}
            
              string sqlQuery3 = "SELECT authors.au_fname,count(titleauthor.title_id) as authorcount FROM authors, titleauthor WHERE authors.au_id = titleauthor.au_id GROUP BY authors.au_fname";
            DataTable dt3 = DataBaseConnection.dataAdapterSelect(sqlQuery3);
            foreach (DataRow dr in dt3.Rows)

>             {dataGridView3.Rows.Add(dr["authorcount"]);}

Reviewer: Format the code.

```c#
private void checkpub_Click(object sender, EventArgs e)
            { string idd = puid.Text; 
            string sqlQuery00 = "select * from  publishers where publishers.pub_id='" + idd + "'";
            DataTable dt00 = DataBaseConnection.dataAdapterSelect(sqlQuery00);

            if (dt00.Rows.Count > 0)
            {foreach (DataRow dr in dt00.Rows)
                { dataGridView4.Rows.Clear();
                  dataGridView4.Rows.Add(dr["pub_id"], dr["pub_name"], dr["city"], dr["state"], dr["country"]);}}
            else
            {MessageBox.Show("This Id is does not exist!");}}
```

### Store tab

The Store's tab will display the discounts available using a left outer join Query.

![Store](/engineering-education/connecting-library-application-to-sql-server-using-windows-forms/Store.png).

```c#
string q5 = " select * from stores left outer join discounts on stores.stor_id = discounts.stor_id ";           
            DataTable q55 = DataBaseConnection.dataAdapterSelect(q5);
            
            foreach (DataRow dr in q55.Rows)
            {
                dataGridView5.Rows.Add(dr["stor_name"], dr["discount"]);
            }

> ```

Reviewer: Format the code

### Conclusion

> In this tutorial we have learned how to connect an application to a database and convert any string to a Query that can be understood by the SQL server, we have also learned how to add edit & delete a book from the database & display the information of the books authors publishers & the discounts of each store. Don't forget to test out the code to fully understand how it works.

Reviewer: Add a `Further reading` section and reference few links which would help the reader gain more knowledge about this topic
